src/stest/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+                .map(|string| File::from(string))
+                .collect()
+        } else {
+            files
+        };
+
+        let files = if self.config.test_contents_of_directories {
+            self.expand_directories(files)
+        } else {
+            files
+        };
+
+        let files = files.into_iter()
+            .filter(|file| self.test(file))
+            .map(|file| self.write(stdout, file))
+            .collect::<Vec<File>>();
+
+        stdout.flush()
+            .unwrap();
+
+        if files.is_empty() {
+            Err(())
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn new(config: Config) -> Self {
+        Self { config }
+    }
+
+    fn expand_directories(&self, files: Vec<File>) -> Vec<File> {
+        files.into_iter()
+            .map(|file| self.expand_directory(file))
+            .flatten()
+            .collect()
+    }
+
+    fn expand_directory(&self, file: File) -> Vec<File> {
+        if file.is_directory() {
+            file.read_directory()
+        } else {
+            vec![file]
+        }
+    }
+
+    /// Test that a file passes all configured tests.
+    ///
+    /// Each test is optional. If a test is disabled, the result is true by default. Only if it's
+    /// enabled do we test the given file. This is expressed with the following boolean logic
+    ///
+    ///   !config.is_test_enabled || test
+    ///
+    /// We combine the result of each individual test with logical AND and return the result,
+    /// inverting it if necessary.
+    fn test(&self, file: &File) -> bool {
+        let passes_all_tests =
+            (!self.config.requires_each_file_is_hidden || file.is_hidden()) &&
+            (!self.config.requires_each_file_is_block_special || file.is_block_special()) &&
+            (!self.config.requires_each_file_is_character_special || file.is_character_special()) &&
+            (!self.config.requires_each_file_is_directory || file.is_directory()) &&
+            (!self.config.requires_each_file_exists || file.exists()) &&
+            (!self.config.requires_each_file_is_file || file.is_file()) &&
+            (!self.config.requires_each_file_has_set_group_id || file.has_set_group_id()) &&
+            (!self.config.requires_each_file_is_symbolic_link || file.is_symbolic_link()) &&
+            (self.optionally_test_if_newer_than_oldest_file(file)) &&
+            (self.optionally_test_if_older_than_newest_file(file)) &&
+            (!self.config.requires_each_file_is_pipe || file.is_pipe()) &&
+            (!self.config.requires_each_file_is_readable || file.is_readable()) &&
+            (!self.config.requires_each_file_has_size_greater_than_zero || file.has_size_greater_than_zero()) &&
+            (!self.config.requires_each_file_has_set_user_id || file.has_set_user_id()) &&
+            (!self.config.requires_each_file_is_writable || file.is_writable()) &&
+            (!self.config.requires_each_file_is_executable || file.is_executable());
```

Comment:
If you make all of these return `Result`, you can just `?` on all of them and have this function also return a `Result<bool,String>`. That will short circuit properly.

Reply:
I like this idea, and I've spent a little bit of time (not as much as I was hoping to, though) trying to push out the "evaluation" of the `Result` type into the outer shell of the app -- in either `main` or `lib::App.run`.

However, it's taking me some time to work around certain issues, such as the fact that `filter` can only accept a `bool` returning function, not a `Result<bool, _>`. I think this means I have to wrap the iteration itself in a `try` or `Result`-like expression to get the `Result` type back. But seems worth it to get a single-layer of error-handling at the top-level.